lib/Dialect/FIRRTL/Transforms/IMDCE.cpp

```diff
@@ -0,0 +1,412 @@
+        continue;
+
+      // Otherwise this is an inuput from it or an inout, mark it as alive.
+      markAlive(portVal);
+    }
+    return;
+  }
+
+  // Otherwise this is a defined module.
+  auto fModule = cast<FModuleOp>(op);
+  markBlockExecutable(fModule.getBody());
+
+  // Ok, it is a normal internal module reference so populate
+  // resultPortToInstanceResultMapping.
+  for (size_t resultNo = 0, e = instance.getNumResults(); resultNo != e;
+       ++resultNo) {
+    auto instancePortVal = instance.getResult(resultNo);
+
+    // Otherwise we have a result from the instance.  We need to forward results
+    // from the body to this instance result's SSA value, so remember it.
+    BlockArgument modulePortVal = fModule.getArgument(resultNo);
+
+    // Mark don't touch ports as alive.
+    if (hasDontTouch(modulePortVal)) {
+      markAlive(modulePortVal);
+      markAlive(instancePortVal);
+    }
+
+    resultPortToInstanceResultMapping[modulePortVal].push_back(instancePortVal);
+  }
+}
+
+void IMDCEPass::markBlockExecutable(Block *block) {
+  if (!executableBlocks.insert(block).second)
+    return; // Already executable.
+
+  for (auto &op : *block) {
+    if (isWireOrRegOrNode(&op))
+      markWireOrRegOrNode(&op);
+    else if (auto instance = dyn_cast<InstanceOp>(op))
+      markInstanceOp(instance);
+    else if (isa<FConnectLike>(op))
+      // Skip connect op.
+      continue;
+    else if (auto mem = dyn_cast<MemOp>(op))
+      markMemOp(mem);
+    else if (!mlir::MemoryEffectOpInterface::hasNoEffect(&op))
+      markUnknownSideEffectOp(&op);
+  }
+}
+
+void IMDCEPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===----- Remove unused ports -----==="
+                          << "\n");
+  auto circuit = getOperation();
+  instanceGraph = &getAnalysis<InstanceGraph>();
+  for (auto module : circuit.getBody()->getOps<FModuleOp>()) {
+    if (module.isPublic()) {
+      markBlockExecutable(module.getBody());
+      // Mark the ports of public modules as alive.
+      for (auto port : module.getBody()->getArguments())
+        markAlive(port);
+    }
+  }
+
+  // If a value changed liveness then propagate liveness through its users and
+  // definition.
+  while (!worklist.empty()) {
+    Value changedVal = worklist.pop_back_val();
+    visitValue(changedVal);
+    for (Operation *user : changedVal.getUsers())
+      visitUser(user);
+  }
+
+  // Rewrite module signatures.
+  for (auto module : circuit.getBody()->getOps<FModuleOp>())
+    rewriteModuleSignature(module);
+
+  // Rewrite module bodies parallelly.
+  mlir::parallelForEach(circuit.getContext(),
+                        circuit.getBody()->getOps<FModuleOp>(),
+                        [&](auto op) { rewriteModuleBody(op); });
+}
+
+void IMDCEPass::visitValue(Value value) {
+  assert(isKnownAlive(value) && "only alive values reach here");
+
+  // Driving input ports propagates the liveness to each instance.
```

Comment:
I don't think we are "driving" an input port here, we are loading from it. Maybe "Requiring an input port propagates.."